### PR TITLE
Update pin_memory.py #116403

### LINKED ISSUE #116403: pin_memory() for a custom Python object
### 🐛 Describe the bug

Please see these lines in the latest main branch.
https://github.com/pytorch/pytorch/blob/f6dfbffb3bb46ada6fe66b5da4f989f9d4d69b3c/torch/utils/data/_utils/pin_memory.py#L76-L77

It show it works if there comes a custom Python object implementing its `pin_memory` member function. However, contrary to the other branches, there is no space to pass `device` such as 

```python
 elif hasattr(data, "pin_memory"): 
     return data.pin_memory(device)
```

It seems valid to add such that capability to it.

### Versions

The latest main branch

### PR DESCRIPTION
**Pull request by Yashovardhan08**
Update pin_memory.py to include specific device, if mentioned

Fixes #116403

### CODE REVIEW

File: torch/utils/data/_utils/pin_memory.py [hunk 1]
```diff
@@ -74,6 +74,6 @@ def pin_memory(data, device=None):
             # The sequence type may not support `__init__(iterable)` (e.g., `range`).
             return [pin_memory(sample, device) for sample in data]
     elif hasattr(data, "pin_memory"):
-        return data.pin_memory()
+        return data.pin_memory(device)
```
Comment (vadimkantorov) [inline torch/utils/data/_utils/pin_memory.py:77]: I wonder if some hacks like checking method signature are needed, as probably for many years implementors assumed and implemented an argument-less variant

Comment (Yashovardhan08) [reply torch/utils/data/_utils/pin_memory.py:77]: Sure, I will try to add those checks

Comment (Yashovardhan08) [reply torch/utils/data/_utils/pin_memory.py]: @vadimkantorov I have added some basic checks, please advise

Comment (vadimkantorov) [reply torch/utils/data/_utils/pin_memory.py]: I'm not maintainer :) We need to wait for some real reviewers. The above is just a consideration that seems reasonable. But I don't have a larger design picture in mind

**Comment by albanD** [comment]
Change sonuds good to me.
I'll let @jbschlosser guide you through the rest of the process since he marked the issue as actionable.

**Comment by vadimkantorov** [comment]
@alband should this check be cached to be not performed at every pinning?

**Comment by Yashovardhan08** [comment]
Will this be reviewed? 
@albanD  @jbschlosser

**Pull request closed without merging**